crates/re_chunk/src/batcher.rs

```diff
@@ -0,0 +1,1588 @@
+    }
+}
+
+impl std::fmt::Debug for BatcherHooks {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let Self {
+            on_insert,
+            on_release,
+        } = self;
+        f.debug_struct("BatcherHooks")
+            .field("on_insert", &on_insert.as_ref().map(|_| "…"))
+            .field("on_release", &on_release)
+            .finish()
+    }
+}
+
+// ---
+
+/// Defines the different thresholds of the associated [`ChunkBatcher`].
+///
+/// See [`Self::default`] and [`Self::from_env`].
+#[derive(Clone, Debug, PartialEq)]
+pub struct ChunkBatcherConfig {
+    /// Duration of the periodic tick.
+    //
+    // NOTE: We use `std::time` directly because this library has to deal with `crossbeam` as well
+    // as std threads, which both expect standard types anyway.
+    //
+    // TODO(cmc): Add support for burst debouncing.
+    pub flush_tick: Duration,
+
+    /// Flush if the accumulated payload has a size in bytes equal or greater than this.
+    ///
+    /// The resulting [`Chunk`] might be larger than `flush_num_bytes`!
+    pub flush_num_bytes: u64,
+
+    /// Flush if the accumulated payload has a number of rows equal or greater than this.
+    pub flush_num_rows: u64,
+
+    /// Split a chunk if it contains >= rows than this threshold and one or more of its timelines are
+    /// unsorted.
+    pub max_chunk_rows_if_unsorted: u64,
+
+    /// Size of the internal channel of commands.
+    ///
+    /// Unbounded if left unspecified.
+    pub max_commands_in_flight: Option<u64>,
+
+    /// Size of the internal channel of [`Chunk`]s.
+    ///
+    /// Unbounded if left unspecified.
+    pub max_chunks_in_flight: Option<u64>,
+
+    /// Callbacks you can install on the [`ChunkBatcher`].
+    pub hooks: BatcherHooks,
+}
+
+impl Default for ChunkBatcherConfig {
+    fn default() -> Self {
+        Self::DEFAULT
+    }
+}
+
+impl ChunkBatcherConfig {
+    /// Default configuration, applicable to most use cases.
+    pub const DEFAULT: Self = Self {
+        flush_tick: Duration::from_millis(8), // We want it fast enough for 60 Hz for real time camera feel
+        flush_num_bytes: 1024 * 1024,         // 1 MiB
+        flush_num_rows: u64::MAX,
+        max_chunk_rows_if_unsorted: 256,
+        max_commands_in_flight: None,
+        max_chunks_in_flight: None,
+        hooks: BatcherHooks::NONE,
+    };
+
+    /// Always flushes ASAP.
+    pub const ALWAYS: Self = Self {
+        flush_tick: Duration::MAX,
+        flush_num_bytes: 0,
+        flush_num_rows: 0,
+        max_chunk_rows_if_unsorted: 256,
+        max_commands_in_flight: None,
+        max_chunks_in_flight: None,
+        hooks: BatcherHooks::NONE,
+    };
+
+    /// Never flushes unless manually told to (or hitting one the builtin invariants).
+    pub const NEVER: Self = Self {
+        flush_tick: Duration::MAX,
+        flush_num_bytes: u64::MAX,
+        flush_num_rows: u64::MAX,
+        max_chunk_rows_if_unsorted: 256,
```

Comment:
The interaction between this and "never" seems a bit odd. Is this also considered one of the built-in invariants?

Reply:
Yes -- in general, only global time and space act as batching thresholds, everything else just splits the result further down into smaller pieces.